lib/fathead/ruby/parse.rb

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+# Holds database column values to be output as tab-separated values.
+class OutputRow
+  HEADERS = [
+    :title,
+    :type,
+    :redirect_title,
+    nil,
+    :categories,
+    nil,
+    :related,
+    nil,
+    :links,
+    :disambiguation_content,
+    :image,
+    :abstract,
+    :url
+  ].freeze
+
+  attr_accessor(*HEADERS.compact)
+
+  def initialize
+    yield self
+    freeze
+  end
+
+  def to_s
+    HEADERS.map { |header| escape public_send header if header } * "\t" << "\n"
+  end
+
+  private
+
+  def escape(text)
+    String(text).gsub(/[\\\n\t]/, '\\' => '\\\\', "\n" => '\n', "\t" => '\t')
+  end
+end
+
+# Abstract description of documentation.
+class Documentation
+  attr_accessor :title, :url, :description, :categories, :related, :alt_titles
+
+  def initialize
+    yield self
+    freeze
+  end
+
+  def abstract
+    intro.chunk(&:name).map do |name, elements|
+      if name == 'pre'
+        elements.map { |e| code_block e.text }.join
+      else
+        elements.map { |e| body_text e.text }.join('<br>')
+      end
+    end.unshift(usage).join
+  end
+
+  def to_s
+    [to_row, to_redirect_rows].join
+  end
+
+  def breaks_category_page?
+    title.include?('[') || abstract.include?('[')
+  end
+
+  private
+
+  def to_row
+    OutputRow.new do |row|
+      row.title = title
+      row.type = 'A' # article
+      row.categories = categories.join("\n")
```

Comment:
@moollaza Literal `\n`s are all escaped as `\\n` on [line 30](https://github.com/duckduckgo/zeroclickinfo-fathead/pull/277/files/db1be20bdd8dd5f2fb470160c6e41808567eae18#diff-e5575cf2178993d658969bb2df779a74R30) when the row is printed.

That’s perhaps not obvious from the way the code is organized, though; I’ll have a think about how I can make it clearer.